File: git-epic-creator/services/code_graph_ingestion_service/src/core/ignore_rules.py
"""Ignore rules for deterministic inventory (Task 05).

Uses gitignore-compatible matching via `pathspec`.
"""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path

import pathspec


# Deterministic defaults. Keep stable; expanding is ok if done deliberately.
DEFAULT_IGNORE_PATTERNS: tuple[str, ...] = (
    # VCS
    ".git/",
    ".hg/",
    ".svn/",
    # Python
    "__pycache__/",
    "*.pyc",
    "*.pyo",
    ".pytest_cache/",
    ".mypy_cache/",
    ".ruff_cache/",
    ".tox/",
    ".nox/",
    ".venv/",
    "venv/",
    # Node
    "node_modules/",
    # Java / build
    "target/",
    "build/",
    "dist/",
    "out/",
    # IDE / OS junk
    ".idea/",
    ".vscode/",
    "*.swp",
    "Thumbs.db",
    "Desktop.ini",
)


@dataclass(frozen=True)
class IgnoreRules:
    spec: pathspec.PathSpec

    def is_ignored(self, repo_rel_posix_path: str) -> bool:
        # PathSpec expects POSIX-style paths.
        return self.spec.match_file(repo_rel_posix_path)


def build_ignore_rules(repo_root: Path, extra_patterns: list[str] | None = None) -> IgnoreRules:
    """Build ignore rules from defaults + optional `.gitignore` + extra patterns."""

    patterns: list[str] = list(DEFAULT_IGNORE_PATTERNS)
    if extra_patterns:
        patterns.extend(extra_patterns)

    gitignore = repo_root / ".gitignore"
    if gitignore.is_file():
        # gitignore is line-based and already in gitwildmatch format.
        patterns.extend(gitignore.read_text(encoding="utf-8", errors="replace").splitlines())

    # Deterministic: PathSpec treats order as given; keep stable ordering.
    spec = pathspec.PathSpec.from_lines("gitwildmatch", patterns)
    return IgnoreRules(spec=spec)

File: git-epic-creator/services/code_graph_ingestion_service/src/core/ignore_rules.rs
use ignore::gitignore::Gitignore;
use ignore::gitignore::GitignoreBuilder;
use std::path::Path;

// Deterministic defaults. Keep stable; expanding is ok if done deliberately.
const DEFAULT_IGNORE_PATTERNS: &[&str] = &[
    // VCS
    ".git/**",
    ".hg/**",
    ".svn/**",
    // Python
    "__pycache__/**",
    "*.pyc",
    "*.pyo",
    ".pytest_cache/**",
    ".mypy_cache/**",
    ".ruff_cache/**",
    ".tox/**",
    ".nox/**",
    ".venv/**",
    "venv/**",
    // Node
    "node_modules/**",
    // Java / build
    "target/**",
    "build/**",
    "dist/**",
    "out/**",
    // IDE / OS junk
    ".idea/**",
    ".vscode/**",
    "*.swp",
    "Thumbs.db",
    "Desktop.ini",
];

#[derive(Debug, Clone)]
pub struct IgnoreRules {
    gitignore: Gitignore,
}

impl IgnoreRules {
    pub fn is_ignored(&self, repo_rel_posix_path: &str) -> bool {
        // The ignore crate expects forward slashes for gitignore semantics.
        // It returns (is_match, maybe_whitelist).
        self.gitignore.matched(repo_rel_posix_path, false).is_ignore()
    }
}

pub fn build_ignore_rules(repo_root: &Path, extra_patterns: Option<&[String]>) -> anyhow::Result<IgnoreRules> {
    let mut b = GitignoreBuilder::new(repo_root);
    for pat in DEFAULT_IGNORE_PATTERNS {
        b.add_line(None, pat)?;
    }
    if let Some(extra) = extra_patterns {
        for pat in extra {
            b.add_line(None, pat)?;
        }
    }

    let gitignore_path = repo_root.join(".gitignore");
    if gitignore_path.is_file() {
        let text = std::fs::read_to_string(&gitignore_path)?;
        for line in text.lines() {
            b.add_line(None, line)?;
        }
    }

    let gi = b.build()?;
    Ok(IgnoreRules {
        gitignore: gi,
    })
}


